cmd/gator/validate/validate.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright © 2021 NAME HERE <EMAIL ADDRESS>
+
+*/
+package validate
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/gator/validate"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+var (
+	errLog *log.Logger
+	outLog *log.Logger
+)
+
+const (
+	examples = `  # Validate a manifest containing Kubernetes objects, Constraint Templates, and Constraints
+  gator validate --filename="manifest.yaml"
+
+  # Validate a directory
+  gator validate --filename="config-and-policies/"
+
+  # Use multiple inputs
+  gator validate --filename="manifest.yaml" --filename="templates-and-constraints/"
+
+  # Receive input from stdin
+  cat manifest.yaml | gator validate
+
+  # Output structured violations data
+  gator validate --filename="manifest.yaml" --json`
+)
+
+var Cmd = &cobra.Command{
+	Use:     "validate",
+	Short:   "validate resources against templates and constraints",
+	Example: examples,
+	Run:     run,
+}
+
+var (
+	flagFilenames []string
+	flagYAML      bool
+	flagJSON      bool
+)
+
+const (
+	flagNameFilename = "filename"
+	flagNameJSON     = "json"
+	flagNameYAML     = "yaml"
+)
+
+func init() {
+	Cmd.Flags().StringArrayVarP(&flagFilenames, flagNameFilename, "f", []string{}, "a file or directory containing kubernetes resources.  Can be specified multiple times.  Cannot be used in tandem with stdin.")
+	Cmd.Flags().BoolVar(&flagYAML, flagNameYAML, false, "print validation information as yaml.")
+	Cmd.Flags().BoolVar(&flagJSON, flagNameJSON, false, "print validation information as json.")
+
+	errLog = log.New(os.Stderr, "", 0)
+	outLog = log.New(os.Stdout, "", 0)
+}
+
+func run(cmd *cobra.Command, args []string) {
+	var unstrucs []*unstructured.Unstructured
+
+	// check if stdin has data
+	stdinfo, err := os.Stdin.Stat()
+	if err != nil {
+		errLog.Fatalf("getting info for stdout: %s", err)
+	}
+
+	// using stdin in combination with flags is not supported
+	if stdinfo.Size() > 0 && len(flagFilenames) > 0 {
+		errLog.Fatalf("stdin cannot be used in combination with %q flag", flagNameFilename)
+	}
+
+	// if no files specified, read from Stdin
+	switch {
```

Comment:
The first branch won't shrink if moved to another messages since it's just get unstrucs, check err, append.

The second branch I've moved into a separate function: `readFiles(filenames []string) ([]*unstructured.Unstructured, error)`

The third part of the switch is just an error call, so can leave that alone as well.